### DISCUSSION TITLE
Update of Compositions

### DISCUSSION THREAD

**Discussion by rafaeldomi** [crossplane/crossplane#3779] [Q&A]
Hello,

I'm using compositions to build network resources in AWS. I'm facing a situation that I dont know if its a bug or expected behavior, or even something that i am doing wrong.
Suppose that I have 2 claim that is in READY status.
When I modify the composition, and do something wrong (like a idSelector with the wrong match), the claims will try to do the changes, and will get in ERROR state. When then corrected version of composition is applied, the claims doesn't get this new version, and still try do apply the "old" composition. Seems like a cache.

But, if I do a new claim all works correctly. Seems like the claims that are in error state have a cached version of the composition.

Even restarting the aws-provider the first 2 claims didnt work. After some time, like more than 30 minutes in one of my tests, the claim works.

Note: besides I using the words "version", I'm not using the CompositionRevision feature.

**Comment by jbw976** [comment]
Good question @rafaeldomi - IIRC by default selectors are only resolved once - once they are set then they will not be resolved again to select a new resource.  Some background on this behavior is in https://github.com/crossplane/crossplane-runtime/issues/250.

High level (using previous wording from @negz that I found):

The intended behaviour is that the selector will be resolved to select a resource and set the reference only if the reference is unset - this is intentional for two reasons:

1. A selector could easily match multiple resources. If we resolved the selector on every reconcile (whether there was already a reference set or not) we'd risk the selected resource 'flapping' - selecting a different resource each time.
1. It seems to be the "idiomatic Kubernetes" approach. Typically (though I don't think always) in Kubernetes a selector sets a reference, and you need to "unset" the reference for the label to be resolved again.

However, from that issue, it looks like there is a policy you can set to influence how the selector works - e.g. this field on subnet VPC ID selector:
https://marketplace.upbound.io/providers/upbound/provider-aws/v0.29.0/resources/ec2.aws.upbound.io/Subnet/v1beta1#doc:spec-forProvider-vpcIdSelector-policy-resolve

Do you think this selector resolution behavior could be what you are seeing, as opposed to the underlying composition not being applied?
